content/posts/model-checking-exactly-once.md

```diff
@@ -0,0 +1,202 @@
+    txId,
+    msg,
+    state, 
+    nextState
+begin
+MainLoop:
+    while TRUE do
+    LockInMsg:
+        await ~ IsEmpty(inputQueue);
+        (...)
+    end while;
+end process;
+end algorithm; *)
+{{< / highlight >}}
+
+The model describes the state using set of variables (lines 13-18). Input and output queues are modelled as sets (no applied ordering) with each message having a unique `id` and `dupId` unique for each duplicate. The business data storage (line 14) is a record that hold sequence of snapshots (all versions of the state including the newest), `ver` field used to model optimistic concurrency on writes and `tx` field needed by the algorithm to commit message processing transactions. Finally, there are two variables modelling the outbox store. 
+
+The system starts with all an empty state except the input queue which contains `MessageCount*DupCount` messages. There are two handler (this is controlled by the `Processes` sequence) that operate in a loop (line 27) processing one message at a time.  
+
+The specification already expresses some non-determinisms as there is not coordination between the handlers. E.g. an execution in which the first handler processes all the messages as well as one in which it processes none of the messages both belong to the model.   
+
+#### Termination
+
+Before checking any other poperties it's good to make sure that the system doesn't get stuck in any of the executions. In our case, this will mean that there are no more messages in the input queue and both handler are waiting in `LockInMsg`. This can be expressed in TLA+ with:
+
+{{< highlight prolog "linenos=inline,hl_lines=,linenostart=1" >}}
+Termination == <>(/\ \A self \in Processes: pc[self] = "LockInMsg"
+                  /\ IsEmpty(inputQueue))
+{{< / highlight >}}
+
+The property states that eventually (`<>` operator) any execution leads to a state in which the `inputQueue` is empty and all processes are in `LockInMsg` label.
+
+#### Message processing
+
+Now let's look at message processing logic:
+
+{{< highlight prolog "linenos=inline,hl_lines=,linenostart=1" >}}
+if outbox[msg.id] = NULL then
+   txId := (msg.id-1)*DupCount + msg.dupId;
+   state.tx := txId ||
+   state.history := <<[msgId |-> msg.id, ver |-> state.ver + 1]>> \o state.history ||
+   state.ver := state.ver + 1;
+StageOutbox:
+   outboxStagging[txId] := [msgId |-> msg.id, ver |-> state.ver];
+StateCommit:
+   CommitState(state);
+OutboxCommit:
+   CommitOutbox(txId);
+StateCleanup:
+   CleanupState(state);
+end if;
+{{< / highlight >}}
+
+First, we check if the `outbox` contains a record for the current message. If there is not track of the message we generate a unique `txId` for this physical messages (line 2) and execute the business logic. The business logic execution is modelled by capturing `msgId` and `state.ver` in the snapshot history. Please not that these operations are modelling in-memory hander state changes - `state` is a variable defined locally. 
```

Comment:
Should be "no trace" here probably, not "no track". And "this physical message" instead of "this physical messages"